crypto/evp/pmeth_lib.c

```diff
@@ -248,33 +264,52 @@ static EVP_PKEY_CTX *int_ctx_new(OPENSSL_CTX *libctx,
     if (e == NULL && keytype != NULL) {
         int legacy = is_legacy_alg(id, keytype);
 
-        if (legacy) {
-            /* This could fail so ignore errors */
+        /* This could fail so ignore errors */
+        if (legacy)
             ERR_set_mark();
-        }
 
         keymgmt = EVP_KEYMGMT_fetch(libctx, keytype, propquery);
-        if (legacy) {
+        if (legacy)
             ERR_pop_to_mark();
-        } else if (keymgmt == NULL) {
-            EVPerr(EVP_F_INT_CTX_NEW, EVP_R_FETCH_FAILED);
-            return NULL;
-        }
+        else if (keymgmt == NULL)
+            return NULL;   /* EVP_KEYMGMT_fetch() recorded an error */
+
+        /*
+         * Chase down the legacy NID, as that might be needed for diverse
+         * purposes, such as ensure that EVP_PKEY_type() can return sensible
+         * values, or that there's a better chance to "downgrade" a key when
+         * needed.  We go through all keymgmt names, because the keytype
+         * that's passed to this function doesn't necessarily translate
+         * directly.
+         * TODO: Remove this when #legacy keys are gone.
+         */
+        if (keymgmt != NULL)
+            id = get_legacy_alg_type_from_keymgmt(keymgmt);
```

Comment:
What if id is already something other than -1? Does this override that? Why?

Reply:
What do you propose I do, then?

Hmmm, I *could* do this in a local variable, and check it against `id` unless that one's -1, and throw an error if they differ (or maybe even have an assertion), and only update `id` if it's -1.  How does that sound?

Reply:
I'd have gone for: check if id is -1, if it is then call this function

Reply:
I ended up doing the more paranoid check.

Reply:
Travis failures indicate that I wasn't entirely wrong to be paranoid...  Maybe it's trivial, maybe not...

Reply:
Fixed